lld/lld-c/LLDAsLibraryC.cpp

```diff
@@ -443,3 +443,199 @@ void LLVMDisposeUndefinedLinkerSymbolsEraVM(char *linkerSymbolNames[],
 }
+
+//----------------------------------------------------------------------------//
+
+/// This function generates a linker script for EVM architecture.
+/// \p memBufs - array of input memory buffers with following structure:
+///
+///   memBufs[0] - deploy object code
+///   memBufs[1] - deplyed object code
+///   --------------------------
+///   memBufs[2] - 1-st sub-contract (final EVM bytecode)
+///   ...
+///   memBufs[N] - N-st sub-contract (final EVM bytecode)
+///
+/// Sub contracts are optional. They should have the same ordering as in
+/// the YUL layout.
+///
+/// \p bufIDs - array of string identifiers of the buffers. IDs correspond
+/// to the object names in the YUL layout.
+///
+/// For example, the YUL object:
+///
+///   |--D_105_deploy --||--D_105_deployed --||-- B_40 --|
+///
+///   __datasize_B_40 = 1384;
+///   SECTIONS {
+///     . = 0;
+///     .text : SUBALIGN(1) {
+///       D_105(.text);
+///       __dataoffset_D_105_deployed = .;
+///       D_105_deployed(.text);
+///       __datasize_D_105_deployed = . - __dataoffset_D_105_deployed;
+///       __dataoffset_B_40 = .;
+///       __datasize_D_105 = __dataoffset_B_40 + __datasize_B_40;
+///       LONG(__dataoffset_D_105_deployed);
+///     }
+///
+/// The dot '.' denotes current location in the resulting file.
+/// The purpose of the script is to define datasize/dataoffset absolute symbols
+/// that reflect the YUL layour.
+static std::string creteEVMLinkerScript(ArrayRef<LLVMMemoryBufferRef> memBufs,
+                                        ArrayRef<const char *> bufIDs) {
+  assert(memBufs.size() == bufIDs.size());
+  StringRef dataSizePrefix("__datasize_");
+  StringRef dataOffsetPrefix("__dataoffset_");
+
+  // Define the script part related to the top-level contract.
+  StringRef topName(bufIDs[0]);
+  StringRef deployed(bufIDs[1]);
+
+  // Contains the linker script part corresponding to the top-level contract.
+  // For the example above, this contains:
+  //   D_105(.text);
+  //   __dataoffset_D_105_deployed = .;
+  //   D_105_deployed(.text);
+  //   __datasize_D_105_deployed = . - __dataoffset_D_105_deployed;
+  Twine topLevel = topName + "(.text);\n" + dataOffsetPrefix + deployed +
+                   " = .;\n" + deployed + "(.text);\n" + dataSizePrefix +
+                   deployed + " = . - " + dataOffsetPrefix + deployed + ";\n";
+
+  // Contains symbols whose values are the sizes of the dependent contracts.
+  // For the example above, this contains:
+  //   __datasize_B_40 = 1384;
+  std::string symDatasizeDeps;
+
+  // Contains symbols whose values are the offsets of the dependent contracts.
+  // For the example above, this contains:
+  //   __dataoffset_B_40 = .;
+  std::string symDataOffsetDeps;
+  if (memBufs.size() > 2) {
+    // Define datasize symbols for the dependent contracts. They start after
+    // {deploy, deployed} pair of the top-level contract,.i.e. at index 2.
+    for (unsigned idx = 2; idx < bufIDs.size(); ++idx)
```

Comment:
Nit: Could we avoid recomputing `bufIDs.size()` here and below.

Reply:
Moved it to a variable.